src/Components/Web.JS/src/Boot.Web.ts

```diff
@@ -9,56 +9,121 @@
 import { startWebAssembly } from './Boot.WebAssembly.Common';
 import { shouldAutoStart } from './BootCommon';
 import { Blazor } from './GlobalExports';
 import { WebStartOptions } from './Platform/WebStartOptions';
 import { attachStreamingRenderingListener } from './Rendering/StreamingRendering';
-import { attachProgressivelyEnhancedNavigationListener, detachProgressivelyEnhancedNavigationListener } from './Services/NavigationEnhancement';
+import { NavigationEnhancementCallbacks, attachProgressivelyEnhancedNavigationListener } from './Services/NavigationEnhancement';
 import { WebAssemblyComponentDescriptor } from './Services/ComponentDescriptorDiscovery';
-import { ServerComponentDescriptor, discoverComponents } from './Services/ComponentDescriptorDiscovery';
+import { ServerComponentDescriptor } from './Services/ComponentDescriptorDiscovery';
+import { DescriptorHandler, attachComponentDescriptorHandler, processComponentDescriptors } from './Rendering/DomMerging/BoundarySync';
 
 let started = false;
+let isPerformingEnhnacedNavigation = false;
 let webStartOptions: Partial<WebStartOptions> | undefined;
+let lastActiveElement: Element | null = null;
 
-async function boot(options?: Partial<WebStartOptions>): Promise<void> {
+function boot(options?: Partial<WebStartOptions>) : Promise<void> {
   if (started) {
     throw new Error('Blazor has already started.');
   }
   started = true;
   webStartOptions = options;
 
-  attachStreamingRenderingListener(options?.ssr);
+  const navigationEnhancementCallbacks: NavigationEnhancementCallbacks = {
+    beforeEnhancedNavigation,
+    beforeDocumentUpdated,
+    afterDocumentUpdated,
+    afterEnhancedNavigation,
+  };
+
+  const descriptorHandler: DescriptorHandler = {
+    onDescriptorAdded,
+  };
+
+  attachComponentDescriptorHandler(descriptorHandler);
+  attachStreamingRenderingListener(options?.ssr, navigationEnhancementCallbacks);
 
   if (!options?.ssr?.disableDomPreservation) {
-    attachProgressivelyEnhancedNavigationListener(activateInteractiveComponents);
+    attachProgressivelyEnhancedNavigationListener(navigationEnhancementCallbacks);
   }
 
-  await activateInteractiveComponents();
+  processComponentDescriptors(document);
+
+  return Promise.resolve();
 }
 
-async function activateInteractiveComponents() {
-  const serverComponents = discoverComponents(document, 'server') as ServerComponentDescriptor[];
-  const webAssemblyComponents = discoverComponents(document, 'webassembly') as WebAssemblyComponentDescriptor[];
+function onDescriptorAdded(descriptor: ServerComponentDescriptor | WebAssemblyComponentDescriptor) {
+  switch (descriptor.type) {
+    case 'server':
+      // Start the circuit now so that it's more likely to be ready by the time interactivity actually starts.
+      startCircuitIfNotStarted();
+      addServerDescriptor(descriptor);
+      break;
+    case 'webassembly':
+      startWebAssemblyIfNotStarted();
+      // TODO: WebAssembly.
+      break;
```

Comment:
I haven't yet implemented the WebAssembly parts of this feature, but I believe those should be much simpler since interop can happen synchronously. This eliminates concerns like:
* What if an SSR component comment gets removed while the interactive component is getting added?
* What if the comment was replaced with a new one that has updated parameter values?
* etc.